### PR TITLE
Added an optional end-of-step callback to NonlinearOptimizer.optimize().

### PR DESCRIPTION
**Pull request by luisenp**
## Motivation and Context
This would be useful for visualization and debugging during optimization.

## How Has This Been Tested
The unit tests check that the callback is called with the right type of data. 

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: theseus/optimizer/nonlinear/nonlinear_optimizer.py [hunk 1]
```diff
@@ -333,13 +342,17 @@ def _optimize_loop(
     # `track_best_solution` keeps a **detached** copy (as in no gradient info)
     # of the best variables found, but it is optional to avoid unnecessary copying
     # if this is not needed
+    #
+    # If `end_step_callback` is passed, it's called at the very end of each optimizer
+    # step, with four input arguments: (optimizer, info, delta, step_idx).
```
Comment (mhmukadam) [inline theseus/optimizer/nonlinear/nonlinear_optimizer.py:347]: each optimizer step is each iteration or each forward call i.e. at the termination condition (max_iters, etc)?

**Review by mhmukadam — APPROVED** [review]

File: theseus/optimizer/nonlinear/nonlinear_optimizer.py (hunk 1 above)
Comment (mhmukadam) [reply theseus/optimizer/nonlinear/nonlinear_optimizer.py:347]: yeah end_iter is clearer that it's at the end of iteration

Comment (luisenp) [reply theseus/optimizer/nonlinear/nonlinear_optimizer.py]: Renamed.

**Review by joeaortiz — APPROVED** [review]
LGTM, thanks for adding this!

**Review by mhmukadam — APPROVED** [review]

**Review by bamos — APPROVED** [review]

**Pull request merged**